megamek/src/megamek/client/bot/princess/HeatMap.java

```diff
@@ -0,0 +1,816 @@
+            trimMap(entityMap);
+        }
+
+    }
+
+    /**
+     * Manually update the last known positions of tracked entities
+     * @param game  current game
+     */
+    public void refreshLastKnownCache (Game game) {
+
+        // Don't bother updating destroyed entities
+        trimLastKnownPositions(game);
+
+        for (int curId : lastPositionCache.keySet()) {
+            Entity curEntity = game.getEntity(curId);
+            if (curEntity != null &&
+                    curEntity.isDeployed() &&
+                    curEntity.getPosition() != null &&
+                    !curEntity.isCarcass() &&
+                    (isFriendlyTeam || curEntity.isVisibleToEnemy() || curEntity.isDetectedByEnemy())) {
+                lastPositionCache.put(curId, curEntity.getPosition());
+            }
+        }
+
+    }
+
+
+    /**
+     * Apply the entity position to the activity and movement trackers
+     * @param tracked  entity to process
+     */
+    private void processEntity (Entity tracked) {
+        Coords position = new Coords(tracked.getPosition().getX(), tracked.getPosition().getY());
+
+        // Get the map adjustments, based on entity BV
+        int mapAdjustment = getTeamWeightAdjustment(tracked);
+
+        // Update the team activity tracker, adding the position if not already present
+        updateTeamActivityMap(position, mapAdjustment);
+
+        // Only update the movement trackers if the entity has moved from it's last known position
+        if (!position.equals(lastPositionCache.get(tracked.getId()))) {
+
+            // Interpolate movement from last known position
+            mapAdjustment = getMovementWeightAdjustment(tracked);
+            List<Coords> path = new ArrayList<>();
+            Coords startPosition = lastPositionCache.get(tracked.getId());
+            if (lastPositionCache.containsKey(tracked.getId())) {
+                path = Coords.intervening(startPosition, position);
+                // Remove the starting point, as it was already counted in the last round
+                path.remove(startPosition);
+            } else {
+                path.add(position);
+            }
+
+            updateTeamMovementMap(path, mapAdjustment);
+
+            // If individual entity tracking is enabled, adjust the entity tracker
+            if (trackIndividualUnits) {
+                updateEntityMap(tracked.getId(), path, mapAdjustment);
+            }
+
+            // Stash this position for movement interpolation in the next round
+            updateLastKnown(tracked.getId(), position);
+        }
+    }
+
+    /**
+     * Update the team activity tracker for a given position
+     * @param position    position to update
+     * @param adjustment  modifier to existing value
+     */
+    private void updateTeamActivityMap(Coords position, int adjustment) {
+        int mapValue = teamActivity.getOrDefault(position, 0) + adjustment;
+        teamActivity.put(position, mapValue);
```

Comment:
This probably needs a valid position check, as it looks like the callers don't do any on-board/of-board validation.